examples/utils/kv_cache_utils.py

```diff
@@ -0,0 +1,562 @@
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Dict, List, Union
+
+import numpy as np
+from onnxruntime import OrtValue
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+    from onnxruntime import IOBinding
+    from torch import Tensor
+
+
+class Cache(ABC):
+    """Abstract class for KV cache management."""
+
+    def __init__(
+        self,
+        past_names: List[str],
+        present_names: List[str],
+        batch_size: int,
+        num_kv_heads: int,
+        head_dim: int,
+        dtype: str = "float32",
+    ):
+        """Initialize the cache.
+
+        :param past_names: List of past key and value names.
+        :param present_names: List of present key and value names.
+            Names must be in the same order as past_names.
+        :param batch_size: Batch size.
+        :param num_kv_heads: Number of key-value heads.
+        :param head_dim: Dimension of each key-value head.
+        :param dtype: Data type of the key-value tensors.
+        """
+        self.past_names = past_names
+        self.present_names = present_names
+        self.batch_size = batch_size
+        self.num_kv_heads = num_kv_heads
+        self.head_dim = head_dim
+        self.dtype = dtype
+
+    @abstractmethod
+    def update(self, present_kvs: List["NDArray"]):
+        """Update the cache with the present key-value tensors.
+
+        :param present_kvs: List of present key-value tensors. This must be past key-value tensors
+            concatenated with the key-value tensors from the current step.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_kv_inputs(self) -> Dict[str, "NDArray"]:
+        """Get the key-value tensors to be used as inputs for the next step."""
+        raise NotImplementedError
+
+
+class DynamicCache(Cache):
+    """Dynamic cache that doesn't have a fixed size for the cache tensors.
+
+    It stores the most recent key-value tensors and uses them as inputs for the next step.
+    """
+
+    def __init__(
+        self,
+        past_names: List[str],
+        present_names: List[str],
+        batch_size: int,
+        num_kv_heads: int,
+        head_dim: int,
+        dtype: str = "float32",
+    ):
+        """Initialize the cache.
+
+        :param past_names: List of past key and value names.
+        :param present_names: List of present key and value names.
+            Names must be in the same order as past_names.
+        :param batch_size: Batch size.
+        :param num_kv_heads: Number of key-value heads.
+        :param head_dim: Dimension of each key-value head.
+        :param dtype: Data type of the key-value tensors.
+        """
+        super().__init__(past_names, present_names, batch_size, num_kv_heads, head_dim, dtype)
+        # cache before prompt processing is empty tensor
+        self.cache = {
+            k: np.zeros((self.batch_size, self.num_kv_heads, 0, self.head_dim), dtype=self.dtype)
```

Comment:
For some models, the shape of past_key/_value is (batch_size, past_seq_len, num_kv_heads, head_dim)

Reply:
For time being, since this is only used for phi2/llama, we can leave it as-is.

Reply:
Oh I wasn't aware about that. Thanks for the heads up. I will take it into account when I make the changes to fully support phi2 also.